### ISSUE TITLE
Bounded  Union TypeVar Weirdness

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#2114]
**Describe the bug**

```python
from typing import Callable, TypeVar, Union

T = TypeVar('T')
IntStr = Union[str, int]
T1 = TypeVar('T1', bound=IntStr)
T2 = TypeVar('T2', bound=IntStr)


def custom_eq(x: IntStr, y: IntStr) -> bool:
    return True


def eq(f: Callable[[T1], T2], x: T1, y: T2) -> bool:
    return custom_eq(f(x), y)

```

**To Reproduce**
Run pylance on that. I got two error messages of,

Argument of type "T2@eq" cannot be assigned to parameter "x" of type "IntStr" in function "custom_eq"
  Type "T2@eq" cannot be assigned to type "IntStr"
    Type "str* | int*" cannot be assigned to type "str"
    Type "str* | int*" cannot be assigned to type "int"
      "int*" is incompatible with "str"
      "str*" is incompatible with "int"

Argument of type "T2@eq" cannot be assigned to parameter "y" of type "IntStr" in function "custom_eq"
  Type "T2@eq" cannot be assigned to type "IntStr"
    Type "str* | int*" cannot be assigned to type "str"
    Type "str* | int*" cannot be assigned to type "int"
      "int*" is incompatible with "str"
      "str*" is incompatible with "int"

**Expected behavior**
No errors like mypy. The code runs fine. 

You could type hint it as 

```python
def eq(f: Callable[[T1], IntStr], x: T1, y: IntStr) -> bool:
    return custom_eq(f(x), y)
```

and it will work, but I'd like the usage of T2 just to require that the callable is consistent with y even though custom_eq is broad enough to not need that.

Union is relevant as swapping it to something else like List[int] bound worked.

**VS Code extension or command-line**
pylance.  v2021.7.6-pre.1

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.158, which I just published. It will also be included in the next release of pylance.
